File: solutions/ruby/saddle-points/1/saddle_points.rb
class Grid
  def self.saddle_points(matrix)
    # 1. Validación de matriz vacía
    return [] if matrix.empty? || matrix[0].empty?

    # 2. Obtenemos las columnas transponiendo las filas
    # En Ruby, transpose funciona sobre arreglos de arreglos
    rows = matrix
    columns = matrix.transpose

    # 3. Pre-calculamos los máximos de cada fila y mínimos de cada columna
    row_maxima = rows.map(&:max)
    col_minima = columns.map(&:min)

    points = []

    # 4. Buscamos las coordenadas que cumplen ambas condiciones
    rows.each_with_index do |row, r_idx|
      row.each_with_index do |value, c_idx|
        if value == row_maxima[r_idx] && value == col_minima[c_idx]
          # Construimos el Hash con el formato esperado por el test
          points << { "row" => r_idx + 1, "column" => c_idx + 1 }
        end
      end
    end

    points
  end
end

File: solutions/typescript/saddle-points/1/saddle-points.ts
type Point = {
  row: number;
  column: number;
};

/**
 * Encuentra todos los puntos de silla (Saddle Points) en una matriz dada.
 * * @param matrix - Matriz bidimensional de números que representa las alturas de los árboles.
 * @returns Point[] - Lista de coordenadas (basadas en índice 1) de los puntos de silla hallados.
 */
export function saddlePoints(matrix: number[][]): Point[] {
  const result: Point[] = [];

  // Casos extremos: matriz vacía o filas vacías
  if (matrix.length === 0 || matrix[0].length === 0) {
    return result;
  }

  const rowCount = matrix.length;
  const colCount = matrix[0].length;

  // 1. Encontrar el valor máximo de cada fila
  const rowMaxs = matrix.map(row => Math.max(...row));

  // 2. Encontrar el valor mínimo de cada columna
  const colMins: number[] = [];
  for (let c = 0; c < colCount; c++) {
    let min = Infinity;
    for (let r = 0; r < rowCount; r++) {
      if (matrix[r][c] < min) {
        min = matrix[r][c];
      }
    }
    colMins.push(min);
  }

  // 3. Evaluar qué coordenadas cumplen ambas condiciones simultáneamente
  for (let r = 0; r < rowCount; r++) {
    for (let c = 0; c < colCount; c++) {
      const currentElement = matrix[r][c];

      // Si es el mayor de su fila y el menor de su columna, es un punto válido
      if (currentElement === rowMaxs[r] && currentElement === colMins[c]) {
        result.push({
          row: r + 1,      // Ajuste a índice basado en 1
          column: c + 1   // Ajuste a índice basado en 1
        });
      }
    }
  }

  return result;
}